yoga/node/Node.h

```diff
@@ -31,6 +31,128 @@ namespace facebook::yoga {
+
+      size_t index() const {
+        return currentNodeIndex_;
+      }
+
+      friend bool operator==(const Iterator& a, const Iterator& b) {
+        return a.node_ == b.node_ && a.childIndex_ == b.childIndex_;
+      };
+
+      friend bool operator!=(const Iterator& a, const Iterator& b) {
+        return a.node_ != b.node_ || a.childIndex_ != b.childIndex_;
+      };
+
+     private:
+      void next() {
+        if (childIndex_ + 1 >= node_->getChildCount()) {
+          // if the current node has no more children, try to backtrack and
+          // visit its successor
+          if (backtrack_.empty()) {
+            // if there are no nodes to backtrack to, the last node has been
+            // visited
+            node_ = nullptr;
+            childIndex_ = -1;
+          } else {
+            // pop and restore the latest backtrack entry
+            const auto back = backtrack_.back();
+            backtrack_.pop_back();
+            node_ = back.first;
+            childIndex_ = back.second;
+
+            // go to the next node
+            next();
+          }
+        } else {
+          // current node has more children to visit, go to next
+          ++childIndex_;
+          // skip all display: contents nodes, possibly going deeper into the
+          // tree
+          skipContentsNodes();
+        }
+      }
+
+      void skipContentsNodes() {
+        // get the node that would be returned from the iterator
+        auto currentNode = node_->getChild(childIndex_);
+        while (currentNode->style().display() == Display::Contents &&
+               currentNode->getChildCount() > 0) {
+          // if it has display: contents set, it shouldn't be returned but its
+          // children should in its place push the current node and child index
+          // so that the current state can be restored when backtracking
+          backtrack_.push_back({node_, childIndex_});
+          // traverse the child
+          node_ = currentNode;
+          childIndex_ = 0;
+
+          // repeat until a node without display: contents is found in the
+          // subtree or a leaf is reached
+          currentNode = currentNode->getChild(childIndex_);
+        }
+
+        // if no node without display: contents was found, try to backtrack
+        if (currentNode->style().display() == Display::Contents) {
+          next();
+        }
+      }
+
+      const Node* node_;
+      size_t childIndex_;
+      size_t currentNodeIndex_{0};
+      Backtrack backtrack_;
+
+      friend LayoutableChildren;
+    };
+
+    LayoutableChildren(const Node* node) : node_(node) {}
+
+    Iterator begin() const {
+      if (node_->getChildCount() > 0) {
+        auto result = Iterator(node_, 0);
+        result.skipContentsNodes();
+        return result;
+      } else {
+        return Iterator(nullptr, -1);
+      }
+    }
+
+    Iterator end() const {
+      return Iterator(nullptr, -1);
+    }
+
+   private:
+    const Node* node_;
+  };
```

Comment:
Does this all work out if we add following below? As I'm sure you ran into, missing or incorrect signatures can lead to cryptic errors within standard library, so it's helpful to have something like this which can show whether iterator is valid implementation.

`static_assert(std::input_iterator<LayoutableChildren::Iterator>);`

Reply:
Done in https://github.com/facebook/yoga/pull/1726/commits/5028c3e1f7692ec14e866fbefced9a4042afdd6e